### ISSUE TITLE
Typescript: Dates Are Not Deserialized

### ISSUE THREAD

**Issue by pixelshaded** [swagger-api/swagger-codegen#2776]
If you have a property in a definition that is a string with date-time format, the typescript generated makes that property a Date type. However, the response in the client is given verbatim. It never looks for date strings and actually transforms them in to Date objects. So the project will build, but fails at runtime if you try to access Date specific methods since it's actually a string.

I imagine this relates to:
https://github.com/swagger-api/swagger-codegen/issues/1951

Since this PR focuses on deserializing dates in the response. I would suggest making those properties strings until deserialization is supported, otherwise the type checking isn't actually accurate.

**Comment by pixelshaded** [comment]
Noticing this for format: byte as well. Sets data type to ByteArray but is not deserializing that from the string.

**Comment by wing328** [comment]
@pixelshaded thanks for reporting the issue. Would you have cycle to port the solution from JS to TS: https://github.com/swagger-api/swagger-codegen/pull/2030 ?

**Comment by pixelshaded** [comment]
Potentially. 

Honestly I've got a workflow where the source swagger spec goes through a processor before going in to codegen. In this processor, I essentially make up for any short comings in codegen - remove format date-time and byte so I get strings as the data type, along with removing enums from definitions because they dont get generated correctly atm.

I may be able to take some time this weekend to start looking at these issues.

**Comment by moanrose** [comment]
I would suggest that the desiralization of date are not handled by swagger codegen, or that it is optional, because the proper place for this would be in the $httpProvider.defaults.transformResponse collection as described here http://aboutcode.net/2013/07/27/json-date-parsing-angularjs.html

**Comment by wing328** [comment]
@moanrose thanks for the suggestion. For other langauges (e.g. C#, Java, Ruby, Python, etc), the auto-generated SDK handles the deserialization of string back into datetime object in the language automatically so doing the same in Angular 1or 2 API client will make the developer experience more consistent.

If it's more common for Angular 1 or 2 developers to handle the deserialization in their own program, then one workaround I can think of is to document the datetime string as `type: string` in the OpenAPI/Swagger spec instead.

(Re-post: https://github.com/swagger-api/swagger-codegen/issues/3105#issuecomment-226740310)

**Comment by gregjacobs** [comment]
@moanrose I agree with @wing328 that the deserialization should happen as part of the generated code. 

The article you linked to simply provides one option for a place to deserialize date objects in Angular code, but it doesn't necessarily mean that it's the "proper" place to deserialize date objects. In fact, from a software engineering perspective, I would argue that `$httpProvider.defaults.transformResponse` is not a good place to put this sort of transformation, as it is tying your code specifically to the Angular 1 framework. This makes upgrading your code to Angular 2, or switching frameworks in general more difficult.

**Comment by janslow** [comment]
There are also the other TypeScript clients (e.g., TypeScript Fetch) which don't have a feature like `$httpProvider.defaults.transformResponse`, so deserializing needs to happen in the generated code.

**Comment by dinvlad** [comment]
So we already define Date fields in swagger.json as
```
"DateTime": {
  "format": "date-time",
  "type": "string"
},
```
However, they do not get deserialized from string and instead are defined to be of Date type in the generated code (at least, for Angular 2). Do we still expect this conversion to happen automatically at some point?

Thanks

**Comment by moanrose** [comment]
What we ended up doing, and what works for us is a combination of the following

We are using JSJoda - LocalDate, LocalDateTime on the client to keep dates and date times seperated.

We have configured Spring Boot to serialize all dateTime to a zoned datetime format in UTC - which get deserialized using Local Time Zone on the client. The Dates gets serialized as ISO 8601 Date

We use regular expressions to substitute properties with matching date formats, to instances of corresponding JSJoda date types - in a recursive fashion.

We have modified both the AbstractTypeScriptClientCodegen, api.mustache, and model.mustache to accomplish this

**Comment by dinvlad** [comment]
FWIW, I opted to pass the date fields through the Date pipe in Angular to display them in local time. With that, the input can either be an ISO string or a Date object. But if we test these with Date objects, technically the test input differs from the runtime input of ISO strings, even though the results look the same. So it's just a matter of being pedantic for now.

But if we haven't used pipes, the manual deserialization to local time (e.g., using [toLocaleString()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString)) would incorrectly assume that we get Date objects, and would pass unit tests while failing in production (unless we catch that in e2e).

**Comment by jamieathans** [comment]
@dinvlad I like Swagger `date-time` translating to typescript compound type `string | Date` and  would be an easy fix.

It would also be easy to write mapping utility functions:

`function toDate(dateOrString: string | Date): Date;`

`function toISODateString(dateOrString: string | Date): string;`

**Comment by wing328** [comment]
@karlvr thanks for sharing the solution. Do you mind submitting a PR so that we can more easily review the fix?

**Comment by macjohnny** [comment]
I suggest to configure the type mapping as suggested in https://github.com/swagger-api/swagger-codegen/issues/5587#issuecomment-368805748

**Comment by Serg046** [comment]
Some minutes ago I was going to fix an issue by @wing328 way [described here](https://github.com/swagger-api/swagger-codegen/pull/2030). However I see 7 typescript templates! They are pretty different and don't have a common shared logic. I don't understand which exactly template should be fixed in scope of the current "bug". Maybe we should extract http request logic to a separate file and reuse it in all templates? But then it's strange to have templates like `typescript-fetch`.

Also it looks like at least one template doesn't have datetime issue: see [typescript-node](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/typescript-node/api.mustache#L110)

I see there is a beautiful common javascript template. It would be cool to have typescript analog.
So what should be the next step?

**Comment by juancarrey** [comment]
If you are using autogenerated-config.json you can add this config so it gets autogenerated to: `startDate: string | Date`

```
"typeMappings": {
    "Date": "string | Date"
  }
```

**Comment by Phyllon** [comment]
Any news on this topic? This bug seems still to exist.
